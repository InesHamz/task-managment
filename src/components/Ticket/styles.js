import { css } from 'styled-components';

export default css`
  background: white;
  padding: 2rem;
  border-radius: 0.5rem;
  cursor: grab;
  display: flex;
  justify-content: space-between;

  &:not(:last-child) {
    margin-bottom: 2rem;
  }
  
  button {
    :hover {
      cursor: pointer;
    }
  }
`;