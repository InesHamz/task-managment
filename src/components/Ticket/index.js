import * as React from 'react';
import PropTypes from 'prop-types';
import 'styled-components/macro';

import { updateEntryLocalStorage } from '../../utils/localStorage';
import styles from './styles';
import { BoardContext } from '../../context/BoardContext';

const Ticket = ({ data, onDragStart }) => {
  const [ticket, setTicket] = React.useState(data);
  const [hasEditModeOn, setHasEditModeOn] = React.useState(false);
  const { keyLocalStorage } = React.useContext(BoardContext);

  function handleEdit(event) {
    event.preventDefault();

    const newTitle = event.target.elements.ticketEdit.value;
    updateEntryLocalStorage(keyLocalStorage, ticket, newTitle);

    setTicket({ ...ticket, title: newTitle });
    setHasEditModeOn(false);
  }

  React.useEffect(() => {
    setTicket(data);
  }, [data]);

  return (
    <div
      css={styles}
      key={ticket.id}
      draggable={!hasEditModeOn}
      onDragStart={e => onDragStart(e, ticket.id)}
    >
      {!hasEditModeOn && ticket && <span title="ticket">{ticket.title}</span>}
      {hasEditModeOn && (
        <form onSubmit={handleEdit}>
          <div>
            <textarea id="ticket-edit-field" name="ticketEdit" type="text" defaultValue={ticket.title} />
          </div>
          <div>
            <button type="edit">Save card</button>
          </div>
        </form>
      )}
      {!hasEditModeOn && <button onClick={() => setHasEditModeOn(true)}>Edit</button>}
    </div>
  );
};

Ticket.defaultProps = {
  onDragStart: () => {},
};

Ticket.propTypes = {
  data: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    column: PropTypes.number,
  }).isRequired,
  onDragStart: PropTypes.func,
};

export default Ticket;
