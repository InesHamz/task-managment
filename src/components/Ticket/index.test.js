import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import 'jest-styled-components';
import BoardContextProvider from '../../context/BoardContext';
import Ticket from '.';

const mockTicket = { id: 1171906, title: 'test', column: 3 };

describe('Ticket', () => {
  test('should match the snapshot', () => {
    const tree = render(
      <BoardContextProvider value={{ keyLocalStorage: 'test-key' }}>
        <Ticket data={mockTicket} />
      </BoardContextProvider>,
    );

    expect(tree.baseElement).toMatchSnapshot();
  });

  test('clicking edit button allows typing in textarea and saving the new text', () => {
    render(
      <BoardContextProvider value={{ keyLocalStorage: 'test-key' }}>
        <Ticket data={mockTicket} />
      </BoardContextProvider>,
    );

    let initialMessage = mockTicket.title;
    const editButton = screen.getByRole('button', { name: /edit/i });
    userEvent.click(editButton);

    const textarea = screen.getByRole('textbox');
    let extraMessage = ' new';
    userEvent.type(textarea, extraMessage);
    expect(textarea).toHaveValue(`${initialMessage}${extraMessage}`);

    const saveButton = screen.getByRole('button', { name: /save card/i });
    userEvent.click(saveButton);

    let finalMessage = screen.getByTitle(/ticket/i);
    expect(finalMessage).toHaveTextContent(`${initialMessage}${extraMessage}`);
  });
});
