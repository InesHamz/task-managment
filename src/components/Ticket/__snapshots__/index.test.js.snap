// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticket should match the snapshot 1`] = `
.c0 {
  background: white;
  padding: 2rem;
  border-radius: 0.5rem;
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0:not(:last-child) {
  margin-bottom: 2rem;
}

.c0 button:hover {
  cursor: pointer;
}

<body>
  <div>
    <div
      class="c0"
      draggable="true"
    >
      <span
        title="ticket"
      >
        test
      </span>
      <button>
        Edit
      </button>
    </div>
  </div>
</body>
`;
